src/cloud-api-adaptor/pkg/podnetwork/common.go

```diff
@@ -4,42 +4,270 @@
+
+	// Get the secondary interface details
+	secIface, secAddrCIDR, secRoute, err := getSecondaryInterfaceDetails(hostNS, hostPrimaryInterface)
+	if err != nil {
+		return fmt.Errorf("failed to get secondary interface details: %w", err)
+	}
+
+	// Move the secondary interface to the pod network namespace
+	err = moveInterfaceToNamespace(hostNS, podNS, secIface, secAddrCIDR, secRoute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func getInterfaceDetails(ns netops.Namespace, iface string) (
+	addrCIDR netip.Prefix, defRoute *netops.Route, err error) {
+
+	link, err := ns.LinkFind(iface)
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to find link %q: %w", iface, err)
+	}
+
+	addrs, err := link.GetAddr()
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get addresses for link %q: %w", link.Name(), err)
+	}
+
+	for _, a := range addrs {
+		if a.Addr().Is4() {
+			addrCIDR = a
+			break
+		}
+	}
+
+	routes, err := ns.RouteList(&netops.Route{Destination: netops.DefaultPrefix})
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get routes on namespace %q: %w", ns.Path(), err)
+	}
+
+	for _, r := range routes {
+		if r.Device == iface && r.Destination.Bits() == 0 && r.Priority < math.MaxInt {
+			defRoute = r
+			break
+		}
+	}
+
+	fmt.Printf("Interface %q IPv4 address %q and default route %v\n", iface, addrCIDR, defRoute)
+
+	return addrCIDR, defRoute, nil
+}
+
+func getSecondaryInterfaceDetails(ns netops.Namespace, primaryInterface string) (
+	secIface string, secAddrCIDR netip.Prefix, secRoute *netops.Route, err error) {
+
+	links, err := ns.LinkList()
+	if err != nil {
+		return "", netip.Prefix{}, nil, fmt.Errorf("failed to list links in namespace %q: %w", ns.Path(), err)
+	}
+
+	for _, link := range links {
+		if isInterfaceFilteredOut(link.Name()) || link.Name() == primaryInterface {
+			continue
+		}
+
+		addr, route, err := getInterfaceDetails(ns, link.Name())
+		if err != nil {
+			return "", netip.Prefix{}, nil, err
+		}
+
+		// The first interface other than the primary having an IPv4 address and a default route
+		// or an IPv4 address in the same subnet as the primary interface with no default route
+		// is considered the secondary interface.
+
+		if !addr.IsValid() {
+			fmt.Printf("Skipping interface %q as it has no valid IPv4 address\n", link.Name())
+			continue
+		}
+
+		if route != nil {
+			fmt.Printf("Secondary interface %q found with IPv4 address %q and route %v\n", link.Name(), addr, route)
+			return link.Name(), addr, route, nil
+		}
+
+		fmt.Printf("Route is nil, checking if %q is in the same subnet as %q\n", addr, primaryInterface)
+		priAddrCIDR, defRoute, err := getInterfaceDetails(ns, primaryInterface)
+		if err != nil {
+			return "", netip.Prefix{}, nil, err
+		}
+		if priAddrCIDR.Bits() == addr.Bits() {
+			secIface = link.Name()
+			secRoute = &netops.Route{
+				Destination: defRoute.Destination,
+				Gateway:     defRoute.Gateway,
```

Comment:
Here you're assuming defRoute is not nil. IIUC, the loop on line 137 may not find a default route, and in that case the method returns nil without an error.

Should we check defRoute != nil before using it here to avoid a possible nil dereference?